### ISSUE TITLE
Rule: Flag package declarations that don't match with the file structure

### ISSUE THREAD

**Issue by vanniktech** [arturbosch/detekt#646]
`io/gitlab/arturbosch/detekt/generator/JCommander.kt ` should have a package declaration of `package io.gitlab.arturbosch.detekt.generator` and not `package io.gitlab.arturbosch.detekt.cli`.

Noticed in #645

**Comment by vanniktech** [comment]
ktlint 0.29.0 is handling this case already. Hence I'm closing this. No need to duplicate the rule.

**Comment by marschwar** [comment]
ktlint actually stopped handling this. See https://github.com/pinterest/ktlint/issues/280#issuecomment-425437625

**Comment by arturbosch** [comment]
We have an absolute and relative path for each KtFile.
Assuming `src/main/kotlin` as the source set we could implement this rule (should be configurable).

First ideas:
```kotlin
internal class PackageMatchesFileStructureSpec : Spek({

    describe("a") {

        val subject by memoized { PackageMatchesFileStructure(Config.empty) }

        it("b") {
            assertThat(subject.lint(Case.Library.path())).hasSize(1)
        }

        it("c") {
            assertThat(subject.lint(Case.CasesFolder.path().resolve("library/Dummy.kt"))).isEmpty()
        }
    }
})
```

```kotlin
class PackageMatchesFileStructure(config: Config) : Rule(config) {

    override val issue = Issue(javaClass.simpleName, Severity.Style, "", Debt.FIVE_MINS)

    override fun visit(root: KtFile) {
        var path = Paths.get(root.absolutePath()).parent.toString()
        path = if (OSUtils.IS_WINDOWS) {
            path.replace("\\\\", ".")
        } else {
            path.replace("/", ".")
        }
        val pkg = root.packageFqName.toString()
        if (!path.endsWith(pkg)) {
            report(
                CodeSmell(issue, Entity.from(root.packageDirective!!),
                    "Package does not match file structure: '$pkg' vs '${root.relativePath()}'")
            )
        }
    }
}
```

**Comment by marschwar** [comment]
Sorry, I forgot to mention this issue when opening #1617. In my opinion this could be closed as well.

**Comment by arturbosch** [comment]
Oh, poor me could not keep up if the notification count, thank you for taking care of this!
